lighthouse-core/gather/drivers/driver.js

```diff
@@ -447,6 +447,59 @@ class Driver {
       storageTypes: 'all',
     });
   }
+
+  /**
+   * Tracks function call usage. Used by captureJSCalls to inject code into the page.
+   * @param {Function} func The function call to track.
+   * @param {Set} set An empty set to populate with stack traces. Should be
+   *     on the global object.
+   * @return {Function} A wrapper around the original function.
+   */
+  static captureUsage(func, set) {
+    const originalFunc = func;
+    const originalPrepareStackTrace = Error.prepareStackTrace;
+
+    return function() {
+      // See v8's Stack Trace API https://github.com/v8/v8/wiki/Stack-Trace-API#customizing-stack-traces
+      Error.prepareStackTrace = function(error, structStackTrace) {
+        const lastCallFrame = structStackTrace[structStackTrace.length - 1];
+        const file = lastCallFrame.getFileName();
+        const line = lastCallFrame.getLineNumber();
+        const col = lastCallFrame.getColumnNumber();
+        return {url: file, line, col}; // return value is e.stack
+      };
+      const e = new Error(`__called ${func.name}__`);
+      set.add(JSON.stringify(e.stack));
+
+      // Restore prepareStackTrace so future errors use v8's formatter and not
+      // our custom one.
+      Error.prepareStackTrace = originalPrepareStackTrace;
+
+      return originalFunc.apply(this, arguments);
+    };
+  }
+
+  /**
+   * Keeps track of calls to a JS function and returns a list of {url, line, col}
+   * of the usage.
+   * @param {string} funcVar The function name to track ('Date.now', 'console.time').
+   * @param {string} variableToPopulate The variable name to populate with stack traces.
+   *     This should unique and on the global object ('window.__dateNowStackTraces').
+   * @return {Function} Call this method when you want results.
+   */
+  captureJSCalls(funcVar, variableToPopulate) {
+    const collectUsage = () => {
+      return this.evaluateAsync(
+          `(__returnResults(Array.from(${variableToPopulate}).map(item => JSON.parse(item))))`);
+    };
+
+    this.evaluateScriptOnLoad(`
```

Comment:
I'm having a hard time parsing this magic. can we split it out a tad more?

I think this is basically what we're doing, eh?
```js
 this.evaluateScriptOnLoad(`
        ${variableToPopulate} = new Set();
        function ${Driver.captureUsage.toString()};
        ${funcVar} = Driver.captureUsage.name(${funcVar}, ${variableToPopulate});
`);
js